eessi/testsuite/tests/apps/espresso/src/madelung.py

```diff
@@ -0,0 +1,152 @@
+
+# initialize system
+system = espressomd.System(box_l=[100., 100., 100.])
+system.time_step = 0.01
+system.cell_system.skin = 0.4
+
+# set MPI Cartesian topology
+node_grid = system.cell_system.node_grid.copy()
+n_cores = int(np.prod(node_grid))
+if args.topology:
+    system.cell_system.node_grid = node_grid = args.topology
+
+# place ions on a cubic lattice
+base_vector = np.array([1., 1., 1.])
+lattice_size = 3 * [2 * args.size]
+if args.weak_scaling:
+    lattice_size = np.multiply(lattice_size, node_grid)
+system.box_l = np.multiply(lattice_size, base_vector)
+for j in range(lattice_size[0]):
+    for k in range(lattice_size[1]):
+        for l in range(lattice_size[2]):
+            _ = system.part.add(pos=np.multiply([j, k, l], base_vector),
+                                q=(-1.)**(j + k + l), fix=3 * [True])
+
+# setup P3M algorithm
+algorithm = espressomd.electrostatics.P3M
+if args.gpu:
+    algorithm = espressomd.electrostatics.P3MGPU
+solver = algorithm(prefactor=1., accuracy=1e-6)
+if (espressomd.version.major(), espressomd.version.minor()) == (4, 2):
+    system.actors.add(solver)
+else:
+    system.electrostatics.solver = solver
+
+
+print("Algorithm executed. \n")
+
+atol_energy = atol_pressure = 1e-12
+atol_forces = 1e-5
+atol_abs_forces = 2e-6
+
+rtol_energy = 5e-6
+rtol_pressure = 2e-5
+rtol_forces = 0.
+rtol_abs_forces = 0.
+# run checks
+forces = np.copy(system.part.all().f)
+energy, p_scalar, p_tensor = get_normalized_values_per_ion(system)
+ref_energy, ref_pressure = get_reference_values_per_ion(base_vector)
+np.testing.assert_allclose(energy, ref_energy, atol=atol_energy, rtol=rtol_energy)
+np.testing.assert_allclose(p_scalar, np.trace(ref_pressure) / 3.,
+                           atol=atol_pressure, rtol=rtol_pressure)
+np.testing.assert_allclose(p_tensor, ref_pressure, atol=atol_pressure, rtol=rtol_pressure)
+np.testing.assert_allclose(forces, 0., atol=atol_forces, rtol=rtol_forces)
+np.testing.assert_allclose(np.median(np.abs(forces)), 0., atol=atol_abs_forces, rtol=rtol_abs_forces)
+
+
+print("Executing sanity checks...\n")
+if (np.all([np.allclose(energy, ref_energy, atol=atol_energy, rtol=rtol_energy),
+        np.allclose(p_scalar, np.trace(ref_pressure) / 3.,
+                           atol=atol_pressure, rtol=rtol_pressure),
+        np.allclose(p_tensor, ref_pressure, atol=atol_pressure, rtol=rtol_pressure),
+        np.allclose(forces, 0., atol=atol_forces, rtol=rtol_forces),
+            np.allclose(np.median(np.abs(forces)), 0., atol=atol_abs_forces, rtol=rtol_abs_forces)])):
+    print("Final convergence met with tolerances: \n\
+            energy: ", atol_energy, "\n\
+            p_scalar: ", atol_pressure, "\n\
+            p_tensor: ", atol_pressure, "\n\
+            forces: ", atol_forces, "\n\
+            abs_forces: ", atol_abs_forces, "\n")
+else:
+    print("At least one parameter did not meet the tolerance, see the log above.\n")
```

Comment:
The sanity checks have actually already executed at lines 111 to 116, and they will interrupt the Python interpreter with an exception if any check fails, so I would suspect the `else` branch is unreachable. I would also recommend against re-expressing the assertions as `np.allclose` in the conditional to avoid redundancy and prevent the risk that the assertions and conditional diverge over time, for example due to changes to tolerance values.

Reply:
You can `print("Success")` or check if Python returned exit code 0.